packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md

```diff
@@ -0,0 +1,98 @@
+# Nimble Table: Handling Cell State when Scrolling
+
+## Problem Statement
+
+Since the Nimble Table uses virtualization, if the table contains more rows than can be shown in the current table viewport, we will re-use the DOM elements / controls within the table cells as the user scrolls through the table. This is handled by the TanStack virtualizer determining what row indexes are visible at the current scroll position, and then we re-bind the table cell views to the new cell state (including the row data). We need to ensure that cell state isn't incorrectly preserved/ applied to the cells once they represent different row values.
+
+Some of the cell state that would incorrectly apply to new rows (once the user scrolls) is:
+
+-   Selected text (which can span multiple rows), for both editable and non-editable cells
+-   Text cursor position (for editable controls in table cells)
+-   Pending input (for editable controls in table cells)
+-   Open popups (action menu or select/combobox cells whose popup is open)
+
+## Links To Relevant Work Items and Reference Material
+
+[#997: Table virtualization: Clear/Commit (and potentially save/restore) row/cell state when scrolling](https://github.com/ni/nimble/issues/997)
+
+## Implementation / Design
```

Comment:
Would it be helpful for us to point to prior art on how various tables handle these use cases?

Reply:
My initial position before reading the spec was "text selection should work as if the table isn't being virtuallized" but after reading your (excellent) in-depth research I am also coming around to the position that it may not be worth doing.

Another good next step would be to give some high-level estimates to Nate of the up-front work and the maintenance burden of this code. Then he can help guide us between the current recommendation in the spec vs more drastic measures like disabling selection on the table entirely.

Reply:
It's worth pointing out that the `nimble-table` text selection issues are present even without virtualization (as I said, I suspect they're more due to the table using shadow DOM).  
[Here is a Storybook link](https://60e89457a987cf003efc0a5b-pffjjtbdsl.chromatic.com/?path=/story/table--table) from before virtualization was added and it shows the same text selection behavior I described.

As Jesse alluded to, disabling text selection for table text cells (`user-select: none`) is another potential option. So the full set of options is probably:
- Do nothing (current behavior). User can select table cell text with difficulty, text remains selected after a scroll (and different text gets selected)
- Disable text selection for table text cells with `user-select: none` (therefore, there's no text selection that would be kept after a scroll)
- Detect and clear text selection on scroll as this HLD proposed.

Reply:
My preference order:
1. disable text selection. It sounds so inconsistent across browsers that I'm worried end users will get frustrated and clients will think it's buggy
2. do nothing.
3. detect and clear text selection on scroll. It seems expensive/fragile and the payoff isn't that high given the challenges selecting text in the first place

I think we should get opinions from the team but if there's much contention I'd say it's Nate's call to decide which option gives the best ROI

Reply:
My vote is do nothing.

I think the basic text selection interaction stuff for single items or a single row is likely fixable (probably the fixes Molly mentioned). I'd be much more generally concerned if not and would want to understand what's going on and how that impacts our controls outside the table too.

I think the selected row scrolling isn't a big issue as discussed in stand-up (not a primary interaction, shouldn't harm functionality, easy to resolve for the user by clicking off to deselect text).

I think disabling all text selection is heavy handed, being able to just quickly copy a cell of data or select some text, right-click, and search google is handy.

I think the programmatic stuff for detecting selected text seems complex and likely fragile. It's known that shadow dom + selection apis have issues and likely to change over time, i.e. from a developer contributing a lot to web component apis in browser: https://twitter.com/rniwa_dev/status/1624193928659607552

Reply:
Molly just posted a PR with the fixes, [here's a Storybook link](https://60e89457a987cf003efc0a5b-miemcebjrx.chromatic.com/?path=/story/table--table&args=data:LargeDataSet).

It fixes the text selection issues in Chrome (doubleclicking any cell text selects that cell) and Safari (you can drag-select text in one cell / multiple cells / a whole row now, and doubleclicking text to select works).  
Firefox still selects a maximum of one cell of text, but it's easy to do so now.

Reply:
With Molly's fixes I'm fine with either "disable text selection" or "do nothing".

Reply:
Updated HLD to state that we won't detect/remove table text selection. Moved the potential implementation for that to Alternatives. Add "Disabling Text Selection" to the Alternatives section too.